### PR TITLE
fixed : subscription payment failed issue | deprecated issues

### PR DESCRIPTION
**Pull request by mayur-vachchewar**
@MSmedal Ready for code review.

### CODE REVIEW

**Comment by MSmedal** [comment]
Please add your changelog entry, update the version number, etc.

File: classes/class-gf-securesubmit.php [hunk 1]
```diff
@@ -1788,13 +1788,13 @@ public function populateCreditCardLastFour($form)
...
             $_POST[ 'input_' . $cc_field['id'] . '_1' ] = 'XXXXXXXXXXXX' . ($response != null
-                ? $response->last_four
+                ? isset($response->details->cardLast4) ? $response->details->cardLast4 : $response->last_four
                 : '');
         }
 
         if(isset($cc_field['id']) && isset($_POST[ 'input_' . $cc_field['id'] . '_4' ])) {
             $_POST[ 'input_' . $cc_field['id'] . '_4' ] = ($response != null
-            ? $response->card_type
+            ? isset($response->details->cardType) ? isset($response->details->cardType) : $response->card_type
```
Comment (MSmedal) [inline classes/class-gf-securesubmit.php]: Nested ternary functions are deprecated. Please re-work these lines

File: classes/class-gf-securesubmit.php [hunk 2]
```diff
@@ -2456,7 +2456,9 @@ private function create_customer($feed, $submission_data, $entry)
         $this->log_debug(__METHOD__ . '(): Customer meta to be created => ' . print_r($acctHolder, 1));
 
         /** @var string $modifier This value helps semi uniqely identify the customer */
-        $modifier = $this->getSecureSubmitJsResponse()->last_four . $this->getSecureSubmitJsResponse()->card_type;
+        $temp_card_last_4 = isset($this->getSecureSubmitJsResponse()->details->cardLast4) ? $this->getSecureSubmitJsResponse()->details->cardLast4 : $this->getSecureSubmitJsResponse()->last_four;
+        $temp_card_type = isset($this->getSecureSubmitJsResponse()->details->cardType) ? $this->getSecureSubmitJsResponse()->details->cardType : $this->getSecureSubmitJsResponse()->card_type;
```
Comment (MSmedal) [inline classes/class-gf-securesubmit.php]: For improved readability, can we break these two lines up? Maybe something like:

```
`$temp_card_last_4 = isset($this->getSecureSubmitJsResponse()->details->cardLast4) ?
            $this->getSecureSubmitJsResponse()->details->cardLast4 :
            $this->getSecureSubmitJsResponse()->last_four;`
```

File: classes/class-gf-securesubmit.php [hunk 3]
```diff
@@ -2496,7 +2498,8 @@ private function create_customer($feed, $submission_data, $entry)
     private function createPaymentMethod($customer)
     {
         $paymentMethod = null;
-        $acct = $this->getSecureSubmitJsResponse()->token_value;
+        $temp_token_value = isset($this->getSecureSubmitJsResponse()->paymentReference) ? $this->getSecureSubmitJsResponse()->paymentReference : $this->getSecureSubmitJsResponse()->token_value;
```
Comment (MSmedal) [inline classes/class-gf-securesubmit.php]: This one too

File: classes/includes/src/Services/Gateway/HpsSoapGatewayService.php [hunk 4]
```diff
@@ -38,7 +38,9 @@ public function doRequest($transaction, $options = array())
         if ($this->_config->developerId != null && $this->_config->developerId != "") {
             $hpsHeader->appendChild($xml->createElement('hps:DeveloperID', $this->_config->developerId));
             $hpsHeader->appendChild($xml->createElement('hps:VersionNbr', $this->_config->versionNumber));
-            $hpsHeader->appendChild($xml->createElement('hps:SiteTrace', $this->_config->siteTrace));
+            if(!empty($this->_config->siteTrace)) {
```
Comment (MSmedal) [inline classes/includes/src/Services/Gateway/HpsSoapGatewayService.php]: Why are we adding this in? I don't think the plugin will ever pass in a value here.

**Review by MSmedal — CHANGES_REQUESTED** [review]

File: classes/includes/src/Services/Gateway/HpsSoapGatewayService.php (hunk 4 above)
Comment (mayur-vachchewar) [reply classes/includes/src/Services/Gateway/HpsSoapGatewayService.php]: I have removed this condition.

**Comment by mayur-vachchewar** [comment]
@MSmedal I have updated the PR changes and changelog.

**Review by MSmedal — CHANGES_REQUESTED** [review]
This seems to correct the functionality issue. The only problem I see with this fix is that it breaks our default styling for the new cardholder name field: 
image

... please correct this so it's in keeping with the other form fields and I think we're good here.

**Review by MSmedal — CHANGES_REQUESTED** [review]
We need for the Cardholder field and its label to appear in the same way as the other fields. With the current changes it looks like this:
image

**Review by MSmedal — APPROVED** [review]

**Pull request merged**